### ISSUE TITLE
Implement DXR

### ISSUE THREAD

**Issue by HansKristian-Work** [HansKristian-Work/vkd3d-proton#154]
TODO list:

- [ ] D3D12_RAYTRACING_GEOMETRY_FLAGS
- [ ] Object libraries
- [ ] Static samplers in local RS
- [ ] AddToStateObject
- [ ] ... and the rest

**Comment by mimattr** [comment]
@HansKristian-Work Will this work also extend to allow/enable DLSS for applicable GPUs now [driver support is officially available on Linux](https://www.nvidia.com/download/driverResults.aspx/160555/en-us#:~:text=Implemented%20support%20for%20NVIDIA%20NGX)?

**Comment by Joshua-Ashton** [comment]
No

**Comment by mimattr** [comment]
> No, DLSS is a separate API that'd need wrapping and it's headers etc are not available publicly.

@Joshua-Ashton Thanks for clarifying though DXR without supporting features like DLSS is pretty pointless for the current generation of titles, even with high-end NVIDIA hardware (e.g. Battlefield V, Control, Cyberpunk, DIRT 5, Ghostrunner, Metro Exodus, Wolfenstein: Youngblood)

Hopefully AMD have a more _"Open"_ alternative in development.

**Comment by mimattr** [comment]
> No, DLSS is a separate API that'd need wrapping and it's headers etc are not available publicly.

@Joshua-Ashton Thanks for clarifying, unfortunately DXR without supporting features like DLSS is of little benefit for most titles due to the general performance overhead (e.g. Battlefield V, Control, Cyberpunk 2077, DIRT 5, Ghostrunner, Metro Exodus, Wolfenstein: Youngblood)

Hopefully AMD have a more _"Open"_ alternative in development.

**Comment by shmerl** [comment]
> Hopefully AMD have a more "Open" alternative in development.

AMD mentioned some open API for upscaling that they are developing to be usable on all GPUs. So we'd probably need to wait until that materializes before such features become practically useful.

**Comment by vityafx** [comment]
> > No, DLSS is a separate API that'd need wrapping and it's headers etc are not available publicly.
> 
> @Joshua-Ashton Thanks for clarifying, unfortunately DXR without supporting features like DLSS is of little benefit for most titles due to the general performance overhead (e.g. Battlefield V, Control, Cyberpunk 2077, DIRT 5, Ghostrunner, Metro Exodus, Wolfenstein: Youngblood)
> 
> Hopefully AMD have a more _"Open"_ alternative in development.

I can quite disagree with this. Without any upscaling technologies used I am playing absolutely fine with ray tracing on RTX 3080. Basically, anything higher and including 3070 is capable of playing ray-traced games very well. I get 120-140 fps on Linux in Q2RTX, and have played already all the games using ray tracing on Linux and it does it quite well. SO this comment makes absolutely no sense to me.

**Comment by mimattr** [comment]
> Without any upscaling technologies used I am playing absolutely fine with ray tracing on RTX 3080.

@vityafx Unfortunately brute-force alone is not a realistic option as most users are not using high-end RTX 3XXX series GPUs for realtime ray tracing:

https://store.steampowered.com/hwsurvey/videocard

**Comment by vityafx** [comment]
> > Without any upscaling technologies used I am playing absolutely fine with ray tracing on RTX 3080.
> 
> 
> 
> @vityafx Unfortunately brute-force alone is not a realistic option as most users are not using high-end RTX 3XXX series GPUs for realtime ray tracing:
> 
> 
> 
> https://store.steampowered.com/hwsurvey/videocard

I still don't see a good reason for not implementing the ray tracing translation without dlss.

**Comment by doitsujin** [comment]
Friendly reminder that AMD supports DXR as well without supporting Nvidia's DLSS.

We have to start somewhere and relying on one vendor's proprietary upsampling solution isn't going to help anybody.

**Comment by vityafx** [comment]
> Friendly reminder that AMD supports DXR as well without supporting Nvidia's DLSS.
> 
> 
> 
> We have to start somewhere and relying on one vendor's proprietary upsampling solution that we have absolutely no control over whatsoever isn't going to help anybody.

Hand seems to have already started, so kudos for him. Monitoring the branch 24/7, just in case I see something I can help with.

**Comment by oscarbg** [comment]
> No, DLSS is a separate API that'd need wrapping and it's headers etc are not available publicly.

on this topic maybe NVIDIA's @liam-middlebrook (that is updating Vulkan Wine support) or other from NV Linux team is interested in doing a DLSS Wine wrapper (will also require translating DLSS D3D12 mode calls to DLSS Vulkan mode calls).. anyway I think will possibly be easier to support DLSS >2.0 titles i.e. that require no special training per game and shipping the "weights" of NN in the driver..

**Comment by vityafx** [comment]
Sad news, DLSS probably won't ever work in Wine, there are multiple reasons. I have done a lot of work for the past few days for it. I won't share the details as this is not a hundred percent chance that I am correct (still have ideas), but Nvidia really tried hard to make this non-working in wine. Once I will run out of ideas completely or will just stuck or something like that, I will share, but until then, everything I share - nvidia can find it out and "close" even these doors which are deeply hidden.

So, I am more and more looking forward to seeing AMD's solution.

**Comment by vityafx** [comment]
Sad news, DLSS probably won't ever work in Wine, there are multiple reasons. I have done a lot of work for the past few days for it. I won't share the details as this is not a hundred percent chance that I am correct (still have ideas), but Nvidia really tried hard to make this non-working in wine. Once I will run out of ideas completely or will just stuck or something like that, I will share, but until then, everything I share - nvidia can find it out and "close" even these doors which are deeply hidden.

So, I am more and more looking forward to seeing AMD's solution.

**Comment by vityafx** [comment]
> > No, DLSS is a separate API that'd need wrapping and it's headers etc are not available publicly.
> 
> on this topic maybe NVIDIA's @liam-middlebrook (that is updating Vulkan Wine support) or other from NV Linux team is interested in doing a DLSS Wine wrapper (will also require translating DLSS D3D12 mode calls to DLSS Vulkan mode calls).. anyway I think will possibly be easier to support DLSS >2.0 titles i.e. that require no special training per game and shipping the "weights" of NN in the driver..

Sad news, DLSS probably won't ever work in Wine, there are multiple reasons. I have done a lot of work for the past few days for it. I won't share the details as this is not a hundred percent chance that I am correct (still have ideas), but Nvidia really tried hard to make this non-working in wine. Once I will run out of ideas completely or will just stuck or something like that, I will share, but until then, everything I share - nvidia can find it out and "close" even these doors which are deeply hidden.

So, I am more and more looking forward to seeing AMD's solution.

**Comment by Sinity** [comment]
> No, DLSS is a separate API that'd need wrapping and it's headers etc are not available publicly.

Probably a dumb question; did somebody try contacting Nvidia and asking them about it? I don't see a reason why they'd need to restrict this particular feature on Linux. Error 43 was at least understandable (product segmentation).

**Comment by vityafx** [comment]
> > No, DLSS is a separate API that'd need wrapping and it's headers etc are not available publicly.
> 
> Probably a dumb question; did somebody try contacting Nvidia and asking them about it? I don't see a reason why they'd need to restrict this particular feature on Linux. Error 43 was at least understandable (product segmentation).

TLDR: vulkan-to-vulkan is not possible at this moment, d3d is still being researched (slowly now).

Well, I did a few times, no one has responded to my questions. Anyway, the latest findings of mine are most likely to end up with nothing, apparently, the ngx on windows is closely tied to nvapi there, and so it requires some undocumented features of nvapi implemented while the reverse engineering is probably not suitable for wine, hence I may do that for myself, but I will not be able to publish this (yet to confirm that); anyway, it will require months of researching of what exactly is required there. Probably, vulkan-to-vulkan can work as it goes a completely different path even in nvapi, but still, Wolfenstein Youngblood goes also a different path, probably due to using an old(?) and undocumented GfeSDK.dll: apparently, it uses this library to check whether the hardware or system is capable of ngx and doesn't even load the ngx library if the GfeSDK reports it shouldn't. That said, Youngblood doesn't even load the ngx at all on Linux, while other games do. Having more games using dlss and vulkan should help, so far there is only one single game that doesn't even load the corresponding libraries on Linux and it is hardly anyone can do something about it, as it uses a closed and probably an old library which I couldn't find any information about (GfeSDK). I assume Youngblood developers do it wrong then, as they should rely on the nvidia-ngx library to detect the os/hardware capabilities and don't do that on their own. What is also possible to do (though I haven't researched that yet) is mocking the GfeSDK.dll: the API seems to have not too many functions, I've mocked all of them just to see the outcome, but still haven't had any progress there, needs more work on that front. Moreover, I do not think this is the best way to go, as again, games should always load the nvngx library and rely on its behaviour, rather than doing something else to enable DLSS.

D3D translation is still being researched by me, but after I fix one thing in nvapi, there goes another one, so slowly going closer to a finish, but the final result is uncertain, I am not sure that even if I let it work (the ngx library), the DLSS will actually work and not segfault and anything else, as we don't have d3d support on Linux but rather translation layers and the DLSS API changes the d3d objects. Even if vkd3d/dxvk's translation is 100% perfect, I myself am not sure whether it will work fine.

P.S. I do not think this issue is the best one to discuss DLSS and anything else except DXR implementation: these two are completely separate things, moreover, the DLSS and all the corresponding changes are not related to vkd3d/dxvk at all as of now.

**Comment by sfjuocekr** [comment]
Does not seem to work with SoTTR.

**Comment by Bananaman** [comment]
I wonder what happens on the machines that "have to run `nvidia-modprobe -u -c=0`".

All that does is:

- Creates `/dev/nvidia0` which everyone already has (check `ls /dev/nvidia*`)
- Sets it up with "UNIFIED MEMORY" model which is only related to CUDA cores (not games).

I literally cannot figure out why this tip to run that useless command keeps cropping up on forums. Perhaps some distro packages a broken nvidia driver that doesn't create the device automatically, that is my only guess.

**Comment by sfjuocekr** [comment]
I'm on Ubuntu 20.10 and /dev/nvidia0 is present.

**Comment by sfjuocekr** [comment]
> I wonder what happens on the machines that so-called "have to" run `nvidia-modprobe -u -c=0`.
> 
> All that does is:
> 
> * Creates `/dev/nvidia0` which everyone already has (check `ls /dev/nvidia*`)
> * Sets it up with "UNIFIED MEMORY" model which is only related to CUDA library (not games).
> 
> I literally cannot figure out why this tip to run that useless command keeps cropping up on forums. Perhaps some distro has a broken nvidia driver that doesn't create the device automatically, that is my only guess.
> 
> Edit: Out of interest, I am on Fedora 35 with NVIDIA 495.46 beta. No, I don't have to run the command.

I'm not entirely sure why this happens, but I've had to run this command whenever I installed the vulkan beta driver manually.

There currently seems to be something wonky going on with the Wrath of the Lich King and DXR, when I turn on DXR the game slows to a crawl even on a 3080!

**Comment by sfjuocekr** [comment]
Today I tried CyberPunk 2077 with DXR and it flatlines:

[vkd3d.txt](https://github.com/HansKristian-Work/vkd3d-proton/files/9354851/vkd3d.txt)

DLSS works fine.

**Comment by acedogblast** [comment]
@sfjuocekr What GPU are you using? The master build of vkd3d-proton works for me with a RTX 3080 using version 1.5.2 of CP2077.

**Comment by sfjuocekr** [comment]
> @sfjuocekr What GPU are you using? The master build of vkd3d-proton works for me with a RTX 3080 using version 1.5.2 of CP2077.

I have a 3080 as well.

**Comment by acedogblast** [comment]
I have 1.6 CP2077 working with DXR. You just need to build vkd3d-proton with the latest code.

**Comment by sfjuocekr** [comment]
I 'll try that!

**Comment by shadywack** [comment]
> I wonder what happens on the machines that so-called "have to" run `nvidia-modprobe -u -c=0`.
> 
> All that does is:
> 
>     * Creates `/dev/nvidia0` which everyone already has (check `ls /dev/nvidia*`)
> 
>     * Sets it up with "UNIFIED MEMORY" model which is only related to CUDA library (not games).
> 
> 
> I literally cannot figure out why this tip to run that useless command keeps cropping up on forums. Perhaps some distro has a broken nvidia driver that doesn't create the device automatically, that is my only guess.
> 
> Edit: Out of interest, I am on Fedora 35 with NVIDIA 495.46 beta. No, I don't have to run the command.

On the new nVidia 520.56.06 drivers, that command no longer needs to be run. Several distros actually had that as a udev rule, while others explicitly loaded nvidia-ivm.ko to get around this issue we've seen. 
Per the new 520 release notes, this issue was resolved by:
`Updated the Vulkan driver so that the following extensions no longer depend on nvidia-uvm.ko being loaded at runtime:

    VK_KHR_acceleration_structure
    VK_KHR_deferred_host_operations
    VK_KHR_ray_query
    VK_KHR_ray_tracing_pipeline
    VK_NV_cuda_kernel_launch
    VK_NV_ray_tracing
    VK_NV_ray_tracing_motion_blur
    VK_NVX_binary_import
    VK_NVX_image_view_handle
`

**Comment by sfjuocekr** [comment]
Hah, I just read that as well in the release notes!

Now we know :D

**Comment by fontivan** [comment]
With the recent changes (version 2.7?) it seems that the RT shadows used in World of Warcraft can now be enabled. They seem to be working as there are slight differences in the shadows between the settings. These examples are from the Classic version of the game but the settings can also be enabled in the retail Dragonflight version.

Disabled:
![Screenshot from 2022-11-08 19-07-26](https://user-images.githubusercontent.com/21060660/200703775-73b45739-7d13-4473-9d1f-8246111d3002.png)

Good:
![Screenshot from 2022-11-08 19-07-18](https://user-images.githubusercontent.com/21060660/200703795-1de8691e-a5f3-4717-9018-de5abc680c23.png)

High:
![Screenshot from 2022-11-08 19-07-06](https://user-images.githubusercontent.com/21060660/200703809-60b21b62-d6dc-48be-9d5d-b91b230778d9.png)

**Comment by HansKristian-Work** [comment]
Mostly feature complete now, modulo bugs.
